### PR TITLE
Fix segfault in fiber management (issue #11094)

### LINKED ISSUE #11094: Domainslib producing segfaults
Domainslib is producing segfaults on parallel programs, possibly as a consequence of #10875

Was looking at the failures produced by sandmark (https://github.com/ocaml-bench/sandmark/issues/304), the same benchmarks did not produce segmentation fault on `4.12.0+domains`. When trying to narrow down the source of the bug, we landed at #10875 since the failures were absent before its merge. Reverting the changes in #10875 also made the failures to disappear (here's the branch: https://github.com/Sudha247/ocaml/tree/mmap_revert). This needs to be debugged further.

Here's a backtrace of the segfault:

```
#0  scan_stack_frames (stack=0x7fff740149d0, stack=0x7fff740149d0, gc_regs=0x0, fdata=0x7fff86ffcb40, 
    f=0x55555560b610 <oldify_one>) at fiber.c:224
#1  caml_scan_stack (f=f@entry=0x55555560b610 <oldify_one>, fdata=fdata@entry=0x7fff86ffcb40, stack=0x7fff740149d0, 
    gc_regs=gc_regs@entry=0x0) at fiber.c:257
#2  0x000055555560b8d2 in oldify_one (st_v=st_v@entry=0x7fff86ffcb40, v=<optimized out>, p=<optimized out>, 
    p@entry=0x7fffb405fd38) at minor_gc.c:269
#3  0x000055555560bc75 in oldify_mopup (st=st@entry=0x7fff86ffcb40, do_ephemerons=do_ephemerons@entry=0)
    at minor_gc.c:393
#4  0x000055555560c2db in caml_empty_minor_heap_promote (domain=domain@entry=0x7fff74000b20, 
    participating_count=participating_count@entry=20, 
    participating=participating@entry=0x555555880240 <stw_request+64>) at minor_gc.c:622
#5  0x000055555560c6f3 in caml_stw_empty_minor_heap_no_major_slice (domain=domain@entry=0x7fff74000b20, 
    participating_count=20, participating=0x555555880240 <stw_request+64>, unused=<optimized out>) at minor_gc.c:688
#6  0x000055555560c7fe in caml_stw_empty_minor_heap (domain=0x7fff74000b20, unused=<optimized out>, 
    participating_count=<optimized out>, participating=<optimized out>) at minor_gc.c:728
#7  0x00005555555f7ecb in stw_handler (domain=0x7fff74000b20) at domain.c:1083
#8  handle_incoming (s=<optimized out>) at domain.c:274
#9  0x00005555555f972c in caml_handle_incoming_interrupts () at domain.c:287
#10 caml_try_run_on_all_domains_with_spin_work (handler=0x55555560c7f0 <caml_stw_empty_minor_heap>, data=0x0, 
    leader_setup=0x55555560b3f0 <caml_empty_minor_heap_setup>, 
    enter_spin_callback=0x55555560b400 <caml_do_opportunistic_major_slice>, enter_spin_data=0x0) at domain.c:1125
#11 0x000055555560c8fd in caml_empty_minor_heaps_once () at minor_gc.c:784
#12 0x00005555555f7b90 in caml_poll_gc_work () at domain.c:1233
#13 0x000055555560fdb9 in caml_process_pending_actions () at signals.c:243
#14 0x00005555556198b5 in caml_garbage_collection () at signals_nat.c:100
#15 <signal handler called>
#16 0x00005555555ab84a in camlLU_decomposition_multicore__get_534 () at test/LU_decomposition_multicore.ml:21
#17 0x00005555555abd7a in camlLU_decomposition_multicore__fun_691 () at test/LU_decomposition_multicore.ml:50
#18 0x00005555555b0988 in camlDomainslib__Task__work_801 () at lib/task.ml:194
#19 0x00005555555af82b in camlDomainslib__Task__do_task_481 () at lib/task.ml:41
```

cc @kayceesrk @Engil @ctk21

### PR DESCRIPTION
**Pull request by jonludlam**
Each domain maintains a cache of previously allocated stacks that's used to speed up creation/termination of fibers. The cache is organised as an array of linked lists indexed by the size of the stack, and a pointer to the specific list to which the stack should be returned is stashed at the point it is first allocated. Since fibers are not tied to the domain in which they are created but can be moved between domains, this currently means that the domain returning a stack to the cache might be returning it to a different domain's cache. The caches aren't meant to be modified concurrently and this is leading to the segfault in issue 11094.

This PR contains a small test for the issue and a fix. The fix is that rather than maintaining a pointer to the cache, we instead maintain the integer index into the array of lists, and so on freeing the stack the domain returns it to its _own_ cache of stacks rather than the cache of the domain that allocated the stack initially.

Co-authored-by: @kayceesrk and @ctk21 

Fixes: #11094

### CODE REVIEW

File: runtime/fiber.c [hunk 1]
```diff
@@ -124,10 +127,12 @@ alloc_size_class_stack_noexc(mlsize_t wosize, struct stack_info** size_bucket,
   CAML_STATIC_ASSERT(sizeof(struct stack_info) % sizeof(value) == 0);
   CAML_STATIC_ASSERT(sizeof(struct stack_handler) % sizeof(value) == 0);
 
-  if (size_bucket != NULL && *size_bucket != NULL) {
-    stack = *size_bucket;
-    *size_bucket = (struct stack_info*)stack->exception_ptr;
-    CAMLassert(stack->size_bucket == stack_cache_bucket(wosize));
+  if (cache_bucket != -1 &&
+      Caml_state->stack_cache != NULL &&
```
Comment (ctk21) [inline runtime/fiber.c]: Isn't this check redundant? (we will never have `Caml_state->stack_cache` uninitialized at this point)
Maybe move to the stronger `CAMLassert(stack_cache != NULL)`?

File: runtime/fiber.c [hunk 2]
```diff
@@ -124,10 +127,12 @@ alloc_size_class_stack_noexc(mlsize_t wosize, struct stack_info** size_bucket,
...
+      Caml_state->stack_cache[cache_bucket] != NULL) {
+    stack = Caml_state->stack_cache[cache_bucket];
```
Comment (ctk21) [inline runtime/fiber.c]: Please can we move `Caml_state->stack_cache` to a local variable on entry to the function (i.e. `struct stack_info** stack_cache = Caml_state->stack_cache`).

On Mac OS (and maybe elsewhere) the TLS variable `Caml_state` involves a function call and the C compiler doesn't always minimise these calls. Hence best to initialize as a local variable.

File: runtime/fiber.c [hunk 3]
```diff
@@ -468,9 +474,10 @@ void caml_free_stack (struct stack_info* stack)
 {
   CAMLnoalloc;
   CAMLassert(stack->magic == 42);
-  if (stack->size_bucket != NULL) {
-    stack->exception_ptr = (void*)(*stack->size_bucket);
-    *stack->size_bucket = stack;
+  if (stack->cache_bucket != -1 &&
+      Caml_state->stack_cache != NULL) {
```
Comment (ctk21) [inline runtime/fiber.c]: Ditto, isn't this check for a NULL `stack_cache` redundant and so can be an assert?

File: runtime/fiber.c [hunk 4]
```diff
@@ -468,9 +474,10 @@ void caml_free_stack (struct stack_info* stack)
...
+    stack->exception_ptr = (void*)(Caml_state->stack_cache[stack->cache_bucket]);
```
Comment (ctk21) [inline runtime/fiber.c]: Are we able to make this a single `Caml_state` access? (i.e. introduce `struct stack_info** stack_cache = Caml_state->stack_cache`)

File: runtime/fiber.c [hunk 5]
```diff
@@ -468,9 +475,11 @@ void caml_free_stack (struct stack_info* stack)
...
   CAMLnoalloc;
   CAMLassert(stack->magic == 42);
-  if (stack->size_bucket != NULL) {
-    stack->exception_ptr = (void*)(*stack->size_bucket);
-    *stack->size_bucket = stack;
+  if (stack->cache_bucket != -1 &&
+      Caml_state->stack_cache != NULL) {
+    stack->exception_ptr =
+      (void*)(Caml_state->stack_cache[stack->cache_bucket]);
+    Caml_state->stack_cache[stack->cache_bucket] = stack;
```
Comment (ctk21) [inline runtime/fiber.c]: Ditto: `struct stack_info** stack_cache = Caml_state->stack_cache`

File: testsuite/tests/parallel/test_issue_11094.ml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  let run_m = Mutex.create () in
+  let enqueue k = with_mutex run_m (fun () -> Queue.push k run_q) in
+  let dequeue () =
+    with_mutex run_m (fun () -> Queue.take_opt run_q) |> Option.iter (fun k -> continue k ())
+  in
+
+  let rec spawn f =
+    (* Effect handler => instantiates fiber *)
+    match_with f () {
+            retc = (fun () -> dequeue ());
```
Comment (ctk21) [inline testsuite/tests/parallel/test_issue_11094.ml]: Is this indentation intentional vs `exnc` and `effc`?

**Review by ctk21 — COMMENTED** [review]
Good fix for a real problem. Very happy to see a test with lots of fibers moving between domains which surfaces this bug. 

Approach looks sound and so is code. However fiber stack allocation and free is very hot for some fiber based programs and we should try to do our best there. Particularly on Mac OS where accessing a TLS variable can result in function calls for every access.

**Review by kayceesrk — APPROVED** [review]
Aside from the comments from @ctk21, this looks good to me.

**Comment by kayceesrk** [comment]
@jonludlam please can you add a Changes entry.

**Comment by jonludlam** [comment]
Review comments all make sense - I believe I have addressed them all now. Thanks!

**Review by sadiqj — APPROVED** [review]
Implementation matches @jonludlam 's description and has addressed all @ctk21 's asks. Thanks!

**Pull request merged**